src/vector/backends/numpy.py

```diff
@@ -301,6 +301,26 @@ def _shape_of(
     return tuple(shape)
 
 
+def _is_type_safe(
+    array: typing.Union[
+        "VectorNumpy2D",
+        "VectorNumpy3D",
+        "VectorNumpy4D",
+        "MomentumNumpy2D",
+        "MomentumNumpy3D",
+        "MomentumNumpy4D",
+        "CoordinatesNumpy",
+    ],
+) -> bool:
+    for i in range(0, len(array.dtype)):  # type: ignore[arg-type]
+        if not issubclass(
+            array.dtype[i].type, (numpy.integer, numpy.floating)
+        ) or issubclass(array.dtype[i].type, numpy.timedelta64):
+            return False
+
+    return True
```

Comment:
The function now takes in a `vector` and checks the `dtype` provided by the user; the cases where the provided `dtype` mismatches with the array entries are handled by `NumPy`.

As a user can only pass in four tuples inside a `dtype` at max, I feel that iterating over them should be fine. Additionally, the `dtype` checks in every `__array_finalize__` method ensure that the `dtype` reaching here does not have a length of 0.

Reply:
Iterating over metadata is definitely fine! Datasets generally break down into small-metadata and big-data, where the metadata provides a description of the data. Since it's by definition small, it's always okay to iterate over metadata, whereas iterating over data is a thing to do carefully (with NumPy, compiled code, sparingly...).